### DISCUSSION TITLE
Custom resource future support

### DISCUSSION THREAD

**Discussion by FikriMilano** [google/android-fhir#1558] [General]
Custom Resource is not created and showing no stacktrace/error.

I was trying to insert a custom resource called LocationHierarchy into the FHIR engine database. I know it's not supported yet and I expect to receive some stacktrace/error, but turns out nothing show up and the custom resource is not inserted as well.

https://github.com/opensrp/fhir-common-utils/blob/main/src/main/java/org/smartregister/model/location/LocationHierarchy.java

@omarismail94 this is what I found out.

**Comment by omarismail94** [comment]
@FikriMilano  I added the test below locally in [DatabaseImplTest](https://github.com/google/android-fhir/blob/master/engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt)  where I create a LocationHierarchy, save it, then read it from the DB and verify it matches the in-memory object I made and it passed:

```
  @Test
  fun insert_customResourceInDb() = runBlocking {

    val locationHierarchy = LocationHierarchy().setAddress(Address().setCity("Quincy")).setId("123")

    database.insert(locationHierarchy)
    testingUtils.assertResourceEquals(
      locationHierarchy,
      database.select(ResourceType.Bundle, "123")
    )


  }
```

I chose `ResourceType.Bundle` in the select statement as that is the type defined in the LocationHierarchy:
https://github.com/opensrp/fhir-common-utils/blob/main/src/main/java/org/smartregister/model/location/LocationHierarchy.java#L66

I also saved the files in the [location directory](https://github.com/opensrp/fhir-common-utils/tree/main/src/main/java/org/smartregister/model/location) to a directory in the SDK test path so they can be instantiated.

I uploaded the work I did to a new branch here: https://github.com/omarismail94/android-fhir/blob/custom_resource/engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt#L101-L110

Can you either clone it to see if it works for you, or share with me an example of a LocationHierarchy that fails to get saved (I know my example was simple)?


CC: @jingtang10

**Reply by FikriMilano** [reply]
@omarismail94 sorry for the delay.

Mine is very simple as well, though it fails to save.
```
suspend fun saveLocationHierarchy() {
    val locationHierarchy = LocationHierarchy().apply { 
      id = "12345"
    }
    fhirEngine.create(locationHierarchy)
}
```

**Reply by omarismail94** [reply]
Hi @FikriMilano,

How are you seeing if the resource fails to save? I am able to save the custom location hierarchy in the test case below:

```
  @Test
  fun `create custom resource and save it`() = runBlocking {
    val location_id = "123"
    val locationHierarchy =
      LocationHierarchy().setAddress(Address().setCity("Quincy")).setId(location_id)

    val ids = fhirEngine.create(locationHierarchy)
    assertThat(ids).containsExactly(location_id)
    testingUtils.assertResourceEquals(
      locationHierarchy,
      fhirEngine.get<LocationHierarchy>(location_id)
    )
  }
```

The test can be found here:
https://github.com/omarismail94/android-fhir/blob/custom_resource/engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt#L60-L72

**Reply by FikriMilano** [reply]
Thanks @omarismail94, that's good news, I will try that out.

Also, there were literally nothing to see (no logs or crashes) when the resource fails to save.

**Reply by omarismail94** [reply]
> Also, there were literally nothing to see (no logs or crashes) when the resource fails to save.

Did you check the [Database Inspector](https://developer.android.com/studio/inspect/database) (if you are using an Emulator) to see if the DB holds any resources? If so, can you upload a screenshot of the view you see when you try to save things?

**Reply by FikriMilano** [reply]
I just tried to save the LocationHierarchy that you wrote, and it was a success :D (thank you)

```
val location_id = "123"
val locationHierarchy =
      LocationHierarchy().setAddress(Address().setCity("Quincy")).setId(location_id)
```

But then I tried to save the LocationHierarchy that I have and it was a fail.
It seems the problems lays in the resource itself (what do you think?).
I will investigate further.

**Reply by FikriMilano** [reply]
And yes I'm using Database Inspector, here's the screenshot (the successful one from your LocationHierarchy).

<img width="1440" alt="Screen Shot 2022-08-25 at 16 48 06" src="https://user-images.githubusercontent.com/62053304/186633345-bd59dc80-baeb-42ac-ac94-e2d6e881544a.png">

**Reply by FikriMilano** [reply]
If it's the LocationHierarchy that I have, the Database Inspector doesn't have that LocationHierarchy resource saved.

**Reply by omarismail94** [reply]
@FikriMilano sorry I was out on holidays for two weeks and am back today. Can you share the code that you use to build the `LocationHierarchy` object that fails to save?
